File: BaekJoon/Basic/카드정렬하기_1715.cpp
//
//  카드정렬하기_1715.cpp
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/02/06.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: 카드정렬하기 1715
// MARK: - Heap 사용 문제풀이

#if 0
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;
typedef long long ll;

int main() {
    ios_base :: sync_with_stdio(0); cin.tie(0);
    ll N, Ans=0, Sum=0; cin>>N;
    priority_queue<ll, vector<ll>, greater<ll>> PQ;
    for(int i=0; i<N; i++) {
        int num; cin>>num;
        PQ.push(num);
    }
    
    while(PQ.size() > 1) {
        Sum = PQ.top();
        PQ.pop();
        Sum += PQ.top();
        PQ.pop();
        Ans += Sum;
        PQ.push(Sum);
    }
    printf("%lld\n",Ans);
    return 0;
}
#endif

File: BaekJoon/Heap/카드정렬하기_1715.swift
//
//  카드정렬하기_1715.swift
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/02/12.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: 카드정렬하기 1715
// MARK: - 이진탐색 활용 문제풀이
/*
import Foundation

let N = Int(readLine()!)!
var PQ = [Int]()
var Ans = 0

extension Array where Element == Int {
    mutating func binarySearchInsert(_ val: Int) {
        if self.isEmpty { self.append(val); return }
        var left = 0
        var right = self.count-1
        while left <= right {
            let mid = (left + right) / 2
            if self[mid] < val {
                right = mid - 1
            } else {
                left = mid + 1
            }
        }

        self.insert(val, at: left)
    }
}

for _ in 0..<N {
    let num = Int(readLine()!)!
    PQ.binarySearchInsert(num)
}

while PQ.count > 1 {
    let first = PQ.last!
    PQ.removeLast()
    let second = PQ.last!
    PQ.removeLast()
    PQ.binarySearchInsert(first+second)
    Ans += first + second
}

print(Ans)
*/
